File: RTC.Console/Exercises/Chapter11.cs
using RTC.Datastructures;
using RTC.Objects;
using RTC.Primitives;
using RTC.Primitives.Patterns;

namespace RTC.Console.Exercises;

public class Chapter11 : IChapterExercise
{
    public void Run()
    {
        var world = new World();

        var floor = new Plane
        {
            Material = new Material(Colour.White)
            {
                Transparency = 0.9d,
                Reflective = 0.9d
            }
        };
        world.Shapes.Add(floor);

        var leftWall = new Plane
        {
            Transformation = Matrix.TranslationMatrix(0, 0, 5) *
                             Matrix.RotationYMatrix(-Math.PI / 4) *
                             Matrix.RotationXMatrix(Math.PI / 2) *
                             Matrix.ScalingMatrix(10, 0.01d, 10),
            Material = new Material(new CheckerPattern(Colour.White, Colour.Black,
                Matrix.ScalingMatrix(0.06d, 0.06d, 0.06d)))
        };
        world.Shapes.Add(leftWall);

        var rightWall = new Plane
        {
            Transformation = Matrix.TranslationMatrix(0, 0, 5) *
                             Matrix.RotationYMatrix(Math.PI / 4) *
                             Matrix.RotationXMatrix(Math.PI / 2) *
                             Matrix.ScalingMatrix(10, 0.01d, 10),
            Material = new Material(new CheckerPattern(Colour.White, Colour.Black,
                Matrix.ScalingMatrix(0.06d, 0.06d, 0.06d)))
        };
        world.Shapes.Add(rightWall);

        var middle = Sphere.GlassSphere(Matrix.TranslationMatrix(0, 2, 0) * Matrix.ScalingMatrix(1.5d, 1.5d, 1.5d));
        middle.Material = middle.Material with { Ambient = 0.01d, Diffuse = 0.4d };
        world.Shapes.Add(middle);

        world.Lights.Add(new PointLight(Colour.White, Vec4.Point(-10, 10, -10)));

        var camera = new Camera(300, 300, Math.PI / 3)
        {
            Transformation = Matrix.ViewTransform(Vec4.Point(0, 2.5d, -4), Vec4.Point(0, 1.5d, 0), Vec4.Vector(0, 1, 0))
        };

        var image = camera.Render(world);
        image.SavePng("Chapter11.png");
    }
}

File: exercises/chapter11.cpp
#include "RayTracerChallenge/objects/patterns.hpp"
#include <RayTracerChallenge/datastructures/matrix.hpp>
#include <RayTracerChallenge/datastructures/vec4.hpp>
#include <RayTracerChallenge/objects/camera.hpp>
#include <RayTracerChallenge/objects/canvas.hpp>
#include <RayTracerChallenge/objects/colour.hpp>
#include <RayTracerChallenge/objects/plane.hpp>
#include <RayTracerChallenge/objects/sphere.hpp>
#include <RayTracerChallenge/objects/world.hpp>
#include <fstream>
#include <numbers>

int main() {
  rtc::World world{};

  auto floor = rtc::plane();
  auto floorMaterial = floor.getMaterial();
  floorMaterial.pattern =
      std::make_unique<rtc::CheckerPattern>(rtc::WHITE, rtc::BLACK);
  floorMaterial.colour = rtc::WHITE;
  floorMaterial.transparency = 0.9f;
  floorMaterial.reflective = 0.9f;
  floor.setMaterial(floorMaterial);
  world.objects.push_back(std::make_unique<rtc::Plane>(floor));

  auto wall = rtc::plane();
  wall.setTransformationMatrix(
      rtc::rotationMatrixX(std::numbers::pi_v<float> / 2) *
      rtc::translationMatrix(0, 10, 20));
  auto wallMaterial = floor.getMaterial();
  wallMaterial.pattern =
      std::make_unique<rtc::RingPattern>(rtc::WHITE, rtc::BLACK);
  wallMaterial.colour = rtc::Colour{0.5, 0.5, 0.5};
  wallMaterial.specular = 0;
  wall.setMaterial(wallMaterial);
  world.objects.push_back(std::make_unique<rtc::Plane>(wall));

  auto middle = rtc::glassSphere();
  middle.setTransformationMatrix(rtc::translationMatrix(0, 2.0f, 0) *
                                 rtc::scalingMatrix(1.5f, 1.5f, 1.5f));
  auto middleMaterial = middle.getMaterial();
  middleMaterial.ambient = 0.01f;
  middleMaterial.diffuse = 0.3f;
  middle.setMaterial(middleMaterial);
  world.objects.push_back(std::make_unique<rtc::Sphere>(middle));

  world.light = rtc::Light{rtc::WHITE, rtc::point(-10, 10, -10)};

  rtc::Camera camera{144, 144, std::numbers::pi_v<float> / 3};
  camera.setTransform(rtc::viewTransform(
      rtc::point(0, 2.5f, -4), rtc::point(0, 1.75f, 0), rtc::vector(0, 1, 0)));

  const rtc::Canvas canvas = camera.render(world);
  std::ofstream file;
  file.open("image.ppm");
  file << canvas.toPPM();
  file.close();
  return 0;
}
